src/Application.php

```diff
@@ -0,0 +1,168 @@
+    }
+
+    public function run(): void
+    {
+        $patterns = array_map(function (string $pattern): string {
+            return '(' . $pattern . ')';
+        }, self::PATTERNS);
+        $regexp = implode('|', $patterns);
+
+        $excludedDirsParams = [];
+        if (count($this->excludedDirs)) {
+            foreach ($this->excludedDirs as $dir) {
+                $excludedDirsParams[] = '-path';
+                $excludedDirsParams[] = $this->path . $dir;
+                $excludedDirsParams[] = '-o';
+            }
+            // last -o
+            array_pop($excludedDirsParams);
+        }
+
+        $grepCommand = [
+            'grep',
+            '-rP',
+            '--line-number',
+            $regexp,
+            '{}',
+            '+',
+        ];
+
+        $cmd = array_merge([
+            'find',
+            $this->path,
+            '(',
+            '-type',
+            'd',
+            '(',
+        ], $excludedDirsParams, [
+            ')',
+            '-prune',
+            ')',
+            '-o',
+            '-type',
+            'f',
+            '-exec',
+        ], $grepCommand);
+
+        $process = new Process($cmd);
+        $process->run();
+
+        if ($process->getExitCode() !== 0) {
+            // patterns not found
+            $this->exitWithSuccess();
+        }
+        $output = $process->getOutput();
+        $matches = $this->parseOutput($output);
+
+        $errorMessages = $this->buildErrorMessages($matches);
+        if (!count($errorMessages)) {
+            $this->exitWithSuccess();
+        }
+
+        throw new Exception(implode("\n", $errorMessages) . "\n");
+    }
+
+    private function parseOutput(string $output): array
+    {
+        $matches = explode("\n", trim($output));
+        array_walk($matches, 'trim');
+        array_walk($matches, function (string &$match): void {
+            $match = explode(':', $match, 3);
+        });
+        return $matches;
+    }
+
+    private function buildErrorMessages(array $matches): array
+    {
+        $errorMessages = [];
+        foreach ($matches as $match) {
+            $filepath = $match[0];
+            $line = $match[1];
+            $lineContents = $match[2];
+            $violationFile = str_replace($this->path, '', $filepath);
+            if (in_array($violationFile, $this->excludedFiles)) {
+                continue;
+            }
+
+            foreach (self::PATTERNS as $patternName => $pattern) {
+                if (preg_match('~' . $pattern . '~', $lineContents)) {
+                    $errorMessages[] = sprintf(
+                        '%s: %s found on line %s: %s',
+                        $violationFile,
+                        $patternName,
+                        $line,
+                        $lineContents
+                    );
+                    break;
+                }
+            }
+        }
+        return $errorMessages;
+    }
+
+    private function parseInput(): array
```

Comment:
neřešil by tohle celý Symfony Console Commands?

Reply:
Důvod je jednoduchý - minimální dependency, aby to šlo použít v composeru - tohle potřebuje process v x.y, ale v projektu máš x.y-1. Tenhle problém se násobí s počtem závislostí. Takže pro běh jsem chtěl závislostí minimum. Ale pokud to poběží odděleně, tak ten problém padá...

Reply:
dependency tam máš už teď - na `find` a `grep`. implementace findu se liší pro centos, alpine nebo os x a nejsou nijak kompatibilní. takže zabalit do pharu a nepoužívat to jako dependency?

Reply:
Jo, koukal jsem jak to dělá phpstan a přišlo mi to netriviální, tak jsem to do první verze nechtěl řešit a chtěl jsem ověřit, že to vůbec dává smysl.